sdv/metadata/single_table.py

```diff
@@ -0,0 +1,356 @@
+        if sdtype == 'categorical':
+            self._validate_categorical(column_name, **kwargs)
+        elif sdtype == 'numerical':
+            self._validate_numerical(column_name, **kwargs)
+        elif sdtype == 'datetime':
+            self._validate_datetime(column_name, **kwargs)
+        elif sdtype == 'text':
+            self._validate_text(column_name, **kwargs)
+
+    def add_column(self, column_name, **kwargs):
+        """Add a column to the ``SingleTableMetadata``.
+
+        Args:
+            column_name (str):
+                The column name to be added.
+
+            kwargs (type):
+                Any additional key word arguments for the column, where ``sdtype`` is required.
+
+        Raises:
+            - ``ValueError`` if the column already exists.
+            - ``ValueError`` if the ``kwargs`` do not contain ``sdtype``.
+            - ``ValueError`` if the column has unexpected values or ``kwargs`` for the given
+              ``sdtype``.
+        """
+        if column_name in self._columns:
+            raise ValueError(
+                f"Column name '{column_name}' already exists. Use 'update_column' "
+                'to update an existing column.'
+            )
+
+        sdtype = kwargs.get('sdtype')
+        if sdtype is None:
+            raise ValueError(f"Please provide a 'sdtype' for column '{column_name}'.")
+
+        self._validate_column(column_name, **kwargs)
+        self._columns[column_name] = deepcopy(kwargs)
+
+    def update_column(self, column_name, **kwargs):
+        """Update an existing column in the ``SingleTableMetadata``.
+
+        Args:
+            column_name (str):
+                The column name to be updated.
+            **kwargs (type):
+                Any key word arguments that describe metadata for the column.
+
+        Raises:
+            - ``ValueError`` if the column doesn't already exist in the ``SingleTableMetadata``.
+            - ``ValueError`` if the column has unexpected values or ``kwargs`` for the current
+              ``sdtype``.
+        """
+        self._validate_column_exists(column_name)
+        _kwargs = deepcopy(kwargs)
+        if 'sdtype' in kwargs:
+            sdtype = kwargs.pop('sdtype')
+        else:
+            sdtype = self._columns[column_name]['sdtype']
+            _kwargs['sdtype'] = sdtype
+
+        self._validate_column(column_name, sdtype, **kwargs)
+        self._columns[column_name] = _kwargs
+
+    def detect_from_dataframe(self, data):
+        """Detect the metadata from a ``pd.DataFrame`` object.
+
+        This method automatically detects the ``sdtypes`` for the given ``pandas.DataFrame``.
+
+        Args:
+            data (pd.DataFrame):
+                ``pandas.DataFrame`` to detect the metadata from.
+        """
+        if self._columns:
+            raise ValueError(
+                'Metadata already exists. Create a new ``SingleTableMetadata`` '
```

Comment:
There are a lot of `SingleTableMetadata` around the code, maybe just crtrl+f for them and substitute with `MultiTableMetadata`.